File: intrinsic/scene/product/object_world/cc/product_utils.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/scene/product/object_world/cc/product_utils.h"

#include <optional>
#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/strings/substitute.h"
#include "google/protobuf/any.pb.h"
#include "google/protobuf/struct.pb.h"
#include "google/protobuf/wrappers.pb.h"
#include "intrinsic/scene/product/proto/product_world_object_data.pb.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/objects/object_world_client.h"
#include "intrinsic/world/objects/object_world_ids.h"
#include "intrinsic/world/objects/world_object.h"

namespace intrinsic {
namespace product {

namespace {
absl::StatusOr<std::optional<intrinsic_proto::product::ProductWorldObjectData>>
GetProductData(const world::WorldObject& object) {
  if (!object.Proto().object_component().user_data().contains(
          kProductUserDataKey)) {
    return std::nullopt;
  }
  const google::protobuf::Any& product_any =
      object.Proto().object_component().user_data().at(kProductUserDataKey);
  intrinsic_proto::product::ProductWorldObjectData product_data;
  if (!product_any.UnpackTo(&product_data)) {
    return absl::FailedPreconditionError(absl::Substitute(
        "Failed to unpack product data from user data. Expect "
        "type [$0] in user data google.protobuf.Any. Got type url [$1] "
        "instead.",
        intrinsic_proto::product::ProductWorldObjectData::descriptor()
            ->full_name(),
        product_any.type_url()));
  }
  return product_data;
}
}  // namespace

absl::StatusOr<bool> IsProduct(const world::WorldObject& object) {
  INTR_ASSIGN_OR_RETURN(auto product_data, GetProductData(object));
  return product_data.has_value();
}

absl::StatusOr<std::optional<std::string>> GetProductName(
    const world::WorldObject& object) {
  INTR_ASSIGN_OR_RETURN(auto product_data, GetProductData(object));
  if (!product_data.has_value()) {
    return std::nullopt;
  }
  return product_data->product_name();
}

absl::StatusOr<std::optional<google::protobuf::Struct>> GetProductMetadata(
    const world::WorldObject& object) {
  INTR_ASSIGN_OR_RETURN(auto product_data, GetProductData(object));
  if (!product_data.has_value()) {
    return std::nullopt;
  }
  return product_data->metadata();
}

absl::Status CreateObjectFromProduct(
    world::ObjectWorldClient& object_world, const WorldObjectName& object_name,
    const world::ObjectWorldClient::CreateObjectOptions& create_object_options,
    const CreateObjectFromProductParams& params) {
  auto options_with_product_data = create_object_options;
  intrinsic_proto::product::ProductWorldObjectData product_data;
  *product_data.mutable_product_name() = params.product_name;
  *product_data.mutable_metadata() = params.product_metadata;
  google::protobuf::Any product_data_any;
  product_data_any.PackFrom(product_data);
  if (options_with_product_data.user_data.contains(kProductUserDataKey)) {
    return absl::InvalidArgumentError(absl::Substitute(
        "CreateObjectOptions already contains reserved product user data kay "
        "'$0'.",
        kProductUserDataKey));
  }

  options_with_product_data.user_data[kProductUserDataKey] = product_data_any;

  INTR_RETURN_IF_ERROR(object_world.CreateObjectFromSceneObject(
      params.scene_object, object_name, options_with_product_data));

  return absl::OkStatus();
}

}  // namespace product
}  // namespace intrinsic

File: intrinsic/scene/product/object_world/python/product_utils.py
# Copyright 2023 Intrinsic Innovation LLC

"""A module for product world objects based on a user_data field convention."""

from typing import Optional

from google.protobuf import any_pb2
from google.protobuf import struct_pb2
from intrinsic.math.python import data_types
from intrinsic.scene.product.proto import product_world_object_data_pb2
from intrinsic.scene.proto.v1 import scene_object_pb2
from intrinsic.world.python import object_world_client
from intrinsic.world.python import object_world_ids
from intrinsic.world.python import object_world_resources


PRODUCT_USER_DATA_KEY = 'FLOWSTATE_product'

def is_product(world_object: object_world_resources.WorldObject) -> bool:
  """Returns whether the given world object is a product object.

  Args:
    world_object: The world object to check.

  Returns:
    True if the world object contains a valid product user data, otherwise
    False.

  Raises:
    ValueError: If the world object contains a invalid type of user data under
    the PRODUCT_USER_DATA_KEY.
  """
  if PRODUCT_USER_DATA_KEY not in world_object.user_data:
    return False
  if not world_object.user_data[PRODUCT_USER_DATA_KEY].Is(
      product_world_object_data_pb2.ProductWorldObjectData.DESCRIPTOR
  ):
    raise ValueError(
        'World object user data does not contain a valid product user data.'
    )
  return True


def get_product_name(
    world_object: object_world_resources.WorldObject,
) -> str | None:
  """Returns the product name of the given product object.

  Args:
    world_object: The world object to get the product name from.

  Returns:
    The product name of the given object contains valid product user data,
    otherwise None.

  Raises:
    ValueError: If the world object contains a invalid type of user data under
    the PRODUCT_USER_DATA_KEY.
  """
  if not is_product(world_object):
    return None
  world_object_data = product_world_object_data_pb2.ProductWorldObjectData()
  if not world_object.user_data[PRODUCT_USER_DATA_KEY].Unpack(
      world_object_data
  ):
    raise ValueError('Failed to unpack ProductWorldObjectData.')
  return world_object_data.product_name


def get_product_metadata(
    world_object: object_world_resources.WorldObject,
) -> struct_pb2.Struct | None:
  """Returns the product metadata of the given product object.

  Args:
    world_object: The world object to get the metadata from.

  Returns:
    The product metadata of the given object contains valid product user data,
    otherwise None.

  Raises:
    ValueError: If the world object contains a invalid type of user data under
    the PRODUCT_USER_DATA_KEY.
  """
  if not is_product(world_object):
    return None
  world_object_data = product_world_object_data_pb2.ProductWorldObjectData()
  if not world_object.user_data[PRODUCT_USER_DATA_KEY].Unpack(
      world_object_data
  ):
    raise ValueError('Failed to unpack ProductWorldObjectData.')
  return world_object_data.metadata


def create_object_from_product(
    world: object_world_client.ObjectWorldClient,
    object_name: object_world_ids.WorldObjectName,
    product_name: str,
    scene_object: scene_object_pb2.SceneObject,
    product_metadata: Optional[struct_pb2.Struct] = None,
    parent: Optional[object_world_resources.WorldObject] = None,
    parent_object_t_created_object: data_types.Pose3 = data_types.Pose3(),
):
  """Creates an objecf from product data.

  Args:
    world: The object world client.
    object_name: The name of the object to create.
    product_name: The name of the product.
    scene_object: The scene object to create.
    product_metadata: The product metadata to set.
    parent: The parent object to set.
    parent_object_t_created_object: The transformation from the parent object to
      the created object.

  Raises:
    object_world_client.CreateObjectError: If create object fails.
  """
  world_object_data = product_world_object_data_pb2.ProductWorldObjectData(
      product_name=product_name, metadata=product_metadata
  )
  world_object_data_any = any_pb2.Any()
  world_object_data_any.Pack(world_object_data)
  user_data = {PRODUCT_USER_DATA_KEY: world_object_data_any}

  world.create_object(
      object_name=object_name,
      geometry_spec=scene_object,
      parent=parent,
      parent_object_t_created_object=parent_object_t_created_object,
      user_data=user_data,
  )

